### ISSUE TITLE
✅️`Option<&T>`; ❌️`&Option<T>`

### ISSUE THREAD

**Issue by Rudxain** [rust-lang/rust-clippy#13054]
### What it does

Title says it all, but it could be extended to `Result` and other `enum`s

### Advantage

https://github.com/helix-editor/helix/pull/11091

### Drawbacks

If a lib has this lint, it could lead to public API breakage. However, it may be worth it, considering this lint could prevent further breakages

### Example

```rust
fn f<T>(a: &Option<T>) {
    let _ = a.as_ref();
}
```

Could be written as:

```rust
fn f<T>(a: Option<&T>) {
    let _ = a;
}
```

**Comment by TornaxO7** [comment]
Also https://www.youtube.com/watch?v=6c7pZYP_iIE shows some other arguments

**Comment by Rudxain** [comment]
I've already linked Logan Smith's video on the Helix PR, but I didn't include it here because I thought it was redundant

**Comment by TornaxO7** [comment]
> I've already linked Logan Smith's video on the Helix PR, but I didn't include it here because I thought it was redundant

Hm... fair, should I remove my message? (A reaction on this message will be enough)

**Comment by Rudxain** [comment]
> should I remove my message?

Please no! It may help with SEO (I'm aware that acronym has bad connotation, but here on GitHub it can be useful to make issues easier to search).

If you want, just use GH "hide" feature. That's what I do to my own comments, sometimes

**Comment by nyurik** [comment]
There is already a [ref_option_ref](https://rust-lang.github.io/rust-clippy/master/index.html#/ref_option_ref) lint that suggests this change:

```rust
// bad
let x: &Option<&u32> = &Some(&0u32);
// good
let x: Option<&u32> = Some(&0u32);
```

I guess it focuses on the internal code, whereas this issue is about fn signature, but if it is expanded to enums and other places, it might make sense to either split them into multiple lints, or rethink the whole approach?

**Comment by nyurik** [comment]
@Rudxain I implemented rough draft of this in #13336 - but it could use some unit tests. If you have some time, could you provide a few to ensure it works well? Thx!

**Comment by Rudxain** [comment]
> it could use some unit tests. If you have some time, could you provide a few to ensure it works well? Thx!

Do you mean snippet examples? Or do I have to locally build the fork and test clippy on some "real" code?
